### PR TITLE
ENH: Add vtkSingleton.h from 3D Slicer

### PR DESCRIPTION
**Pull request by Sunderlandkyl**
Includes macros for creating singleton vtkObject instances.

### CODE REVIEW

**Comment by Sunderlandkyl** [comment]
Encountered some issues. Converting to draft until they're resolved.

**Comment by Sunderlandkyl** [comment]
Fixed. Needed to add ```Self::Instance->InitializeObjectBase();``` after the instance was created.

**Review by lassoan — COMMENTED** [review]
It looks good, just make sure it follows vtkAddon source code conventions (header, comment style, etc).

**Comment by Sunderlandkyl** [comment]
Updated the header description and added a comment about ctkSingleton.

**Comment by Sunderlandkyl** [comment]
After discussion, we will add a test to this PR to test + give an example implementation of a singleton class using the macros.

File: Testing/vtkMySingletonClass.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+protected:
+  vtkMySingletonClass() = default;
+  ~vtkMySingletonClass() override = default;
+
+  int Value{ 0 };
+
+  VTK_SINGLETON_DECLARE(vtkMySingletonClass);
+};
+
+VTK_SINGLETON_DECLARE_INITIALIZER(, vtkMySingletonClass);
```
Comment (lassoan) [inline Testing/vtkMySingletonClass.h:45]:
```suggestion
VTK_SINGLETON_DECLARE_INITIALIZER(/*add VTK_..._EXPORT directive here if the singleton is accessible outside the library*/, vtkMySingletonClass);
```

File: Testing/vtkMySingletonClass.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+==============================================================================*/
+
+// VTK includes
+#include <vtkObject.h>
+
+// vtkAddon includes
+#include <vtkSingleton.h>
+
+/// \brief Singleton VTK object containing a single integer.
```
Comment (lassoan) [inline Testing/vtkMySingletonClass.h:24]:
```suggestion
/// \brief Simple example VTK singleton class that stores an integer.
```

**Review by lassoan — COMMENTED** [review]
Looks great, thank you! I just added a few minor suggestions (no functional change, just documentation).

**Review by lassoan — APPROVED** [review]

**Comment by lassoan** [comment]
@Sunderlandkyl I've merged this, so you can now update vtkAddon in Slicer (and remove the vtkSingleton implementation from Slicer).

**Pull request merged**